plugins/steelseries/fu-steelseries-sonic.c

```diff
@@ -0,0 +1,1238 @@
+	chip = STEELSERIES_SONIC_CHIP_NORDIC;
+	if (!fu_steelseries_sonic_write_chip(device,
+					     chip,
+					     firmware,
+					     fu_progress_get_child(progress),
+					     flags,
+					     error))
+		return FALSE;
+	fu_progress_step_done(progress);
+	if (!fu_steelseries_sonic_verify_chip(device,
+					      chip,
+					      firmware,
+					      fu_progress_get_child(progress),
+					      error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* holtek */
+	chip = STEELSERIES_SONIC_CHIP_HOLTEK;
+	if (!fu_steelseries_sonic_write_chip(device,
+					     STEELSERIES_SONIC_CHIP_HOLTEK,
+					     firmware,
+					     fu_progress_get_child(progress),
+					     flags,
+					     error))
+		return FALSE;
+	fu_progress_step_done(progress);
+	if (!fu_steelseries_sonic_verify_chip(device,
+					      chip,
+					      firmware,
+					      fu_progress_get_child(progress),
+					      error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_steelseries_sonic_prepare_firmware_chip(FuDevice *device,
+					   SteelseriesSonicChip chip,
+					   GBytes *fw,
+					   gsize offset,
+					   gsize size,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	guint32 checksum_tmp;
+	guint32 checksum;
+	g_autoptr(FuFirmware) firmware = NULL;
+	g_autoptr(GBytes) blob = NULL;
+
+	blob = fu_common_bytes_new_offset(fw, offset, size, error);
+	if (blob == NULL)
+		return NULL;
+	if (!fu_common_read_uint32_safe(g_bytes_get_data(blob, NULL),
+					g_bytes_get_size(blob),
+					g_bytes_get_size(blob) - sizeof(checksum),
+					&checksum,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+	checksum_tmp = fu_common_crc32(g_bytes_get_data(blob, NULL),
+				       g_bytes_get_size(blob) - sizeof(checksum_tmp));
+	checksum_tmp = ~checksum_tmp;
+	if (checksum_tmp != checksum) {
+		if ((flags & FWUPD_INSTALL_FLAG_IGNORE_CHECKSUM) == 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "checksum mismatch for chip %u, got 0x%08x, expected 0x%08x",
+				    chip,
+				    checksum_tmp,
+				    checksum);
+			return FALSE;
+		}
+		g_debug("ignoring checksum mismatch, got 0x%08x, expected 0x%08x",
+			checksum_tmp,
+			checksum);
+	}
+
+	firmware = fu_firmware_new_from_bytes(blob);
+	fu_firmware_set_id(firmware, STEELSERIES_SONIC_FIRMWARE_ID[chip]);
+	fu_firmware_add_flag(firmware, FU_FIRMWARE_FLAG_HAS_CHECKSUM);
+
+	/* success */
+	return g_steal_pointer(&firmware);
+}
+
+static FuFirmware *
+fu_steelseries_sonic_prepare_firmware(FuDevice *device,
+				      GBytes *fw,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuFirmware *firmware_chip;
+	SteelseriesSonicChip chip;
+	gsize offset;
+	gsize size;
+	g_autoptr(FuFirmware) firmware = fu_firmware_new();
+	g_autoptr(FuFirmware) firmware_nordic = NULL;
+	g_autoptr(FuFirmware) firmware_holtek = NULL;
+	g_autoptr(FuFirmware) firmware_mouse = NULL;
+
+	/* TODO: Use FuArchive */
```

Comment:
still the plan?

Reply:
Oh, I would say yes.

I will replace the concatenated blob by a zip archive, and remove the `_dump_firmware()` because there is no write support yet for archive, and because it is wrong as it dumps blobs from different chips.

It should have been `_read_firmware()` right?

Reply:
> It should have been _read_firmware() right?

Ahh yes, indeed.